File: py/test_bands.py
"""Generates composite PNGs of band generation for pixdiff vs test_bands.rs."""

import numpy as np
from PIL import Image
from bands import generate_bands, growth_to_thickness

PPI = (600.0, 600.0)


def make_composite(seed: np.ndarray, bands_list, path: str) -> None:
    """Composite seed + bands into one RGBA PNG.
    seed pixels → black opaque; band-k pixels → white/black opaque; else transparent.
    Black/white in linear space = sRGB 0/255 exactly; no floating point needed.
    """
    h, w = seed.shape
    rgba = np.zeros((h, w, 4), dtype=np.uint8)
    # Seed: black opaque
    rgba[seed, :] = [0, 0, 0, 255]
    # Bands: innermost first = drawn last (but non-overlapping so order doesn't matter)
    for band in bands_list:
        v = 255 if band.white else 0
        rgba[band.mask, :3] = v
        rgba[band.mask, 3]  = 255
    Image.fromarray(rgba, "RGBA").save(path, dpi=PPI)
    print(f"wrote {path}  bands={len(bands_list)}")


def dot_seed(size: int) -> np.ndarray:
    s = np.zeros((size, size), dtype=bool)
    s[size // 2, size // 2] = True
    return s


def diag_seed(size: int) -> np.ndarray:
    s = np.zeros((size, size), dtype=bool)
    for i in range(size):
        s[i, i] = True
    return s


S = 64
C = 256  # confirming run

# dot, growth=30 (thickness=13)
t = growth_to_thickness(30)
seed = dot_seed(S)
bs   = generate_bands(seed, t, 90.0)
make_composite(seed, bs, "test_bands_dot_g30.png")

# dot, growth=10 (thickness=5)
t    = growth_to_thickness(10)
bs   = generate_bands(seed, t, 90.0)
make_composite(seed, bs, "test_bands_dot_g10.png")

# dot, growth=50 (thickness=20)
t    = growth_to_thickness(50)
bs   = generate_bands(seed, t, 90.0)
make_composite(seed, bs, "test_bands_dot_g50.png")

# diagonal line, growth=30
t    = growth_to_thickness(30)
seed = diag_seed(S)
bs   = generate_bands(seed, t, 90.0)
make_composite(seed, bs, "test_bands_diag_g30.png")

# confirming run 256x256 dot, growth=30
t    = growth_to_thickness(30)
seed = dot_seed(C)
bs   = generate_bands(seed, t, 80.0)
make_composite(seed, bs, "test_bands_confirm256.png")

File: rs/src/bin/test_bands.rs
//! Generates composite PNGs of band generation for pixdiff vs py/test_bands.py.

use basecoat::bands::{generate_bands, growth_to_thickness};
use png::{BitDepth, ColorType, Encoder, Unit};
use std::fs::File;
use std::io::BufWriter;

const PPM: u32 = 23622;

fn write_composite(seed: &[bool], bands: &[basecoat::bands::Band], w: u32, h: u32, path: &str) {
    let mut rgba = vec![0u8; (w * h * 4) as usize];

    // Seed: black opaque
    for (i, &s) in seed.iter().enumerate() {
        if s {
            let base = i * 4;
            rgba[base    ] = 0;
            rgba[base + 1] = 0;
            rgba[base + 2] = 0;
            rgba[base + 3] = 255;
        }
    }

    // Bands (non-overlapping, any draw order is fine)
    for band in bands {
        let v = if band.white { 255u8 } else { 0u8 };
        for (i, &m) in band.mask.iter().enumerate() {
            if m {
                let base = i * 4;
                rgba[base    ] = v;
                rgba[base + 1] = v;
                rgba[base + 2] = v;
                rgba[base + 3] = 255;
            }
        }
    }

    let file = File::create(path).expect("create");
    let bw = BufWriter::new(file);
    let mut enc = Encoder::new(bw, w, h);
    enc.set_color(ColorType::Rgba);
    enc.set_depth(BitDepth::Eight);
    enc.set_pixel_dims(Some(png::PixelDimensions { xppu: PPM, yppu: PPM, unit: Unit::Meter }));
    let mut writer = enc.write_header().expect("header");
    writer.write_image_data(&rgba).expect("pixels");
    println!("wrote {path}  bands={}", bands.len());
}

fn dot_seed(size: usize) -> Vec<bool> {
    let mut s = vec![false; size * size];
    s[(size / 2) * size + (size / 2)] = true;
    s
}

fn diag_seed(size: usize) -> Vec<bool> {
    let mut s = vec![false; size * size];
    for i in 0..size { s[i * size + i] = true; }
    s
}

fn run(seed: &[bool], w: usize, h: usize, growth: f32, fill: f32, path: &str) {
    let t     = growth_to_thickness(growth);
    let bands = generate_bands(seed, w, h, t, fill);
    write_composite(seed, &bands, w as u32, h as u32, path);
}

fn main() {
    let s = 64usize;
    let c = 256usize;

    run(&dot_seed(s),  s, s,  30.0, 90.0, "test_bands_dot_g30.png");
    run(&dot_seed(s),  s, s,  10.0, 90.0, "test_bands_dot_g10.png");
    run(&dot_seed(s),  s, s,  50.0, 90.0, "test_bands_dot_g50.png");
    run(&diag_seed(s), s, s,  30.0, 90.0, "test_bands_diag_g30.png");
    run(&dot_seed(c),  c, c,  30.0, 80.0, "test_bands_confirm256.png");
}
